plugins/devlink/fu-devlink-netlink.c

```diff
@@ -0,0 +1,624 @@
+	if (nlg->is_emulated || nlg->save_events)
+		event_id = fu_devlink_netlink_get_event_id(nlg, nlh);
+
+	if (nlg->is_emulated) {
+		event = fu_device_load_event(nlg->device, event_id, error);
+		if (event == NULL)
+			return FALSE;
+	}
+
+	if (nlg->save_events)
+		event = fu_device_save_event(nlg->device, event_id);
+
+	do {
+		if (nlg->is_emulated) {
+			const guint8 *response_buf;
+			gsize response_len;
+			g_autofree gchar *response_key = g_strdup_printf("ResponseData%u", i++);
+			g_autoptr(GBytes) response_data =
+			    fu_device_event_get_bytes(event, response_key, NULL);
+
+			if (response_data == NULL)
+				return TRUE;
+
+			response_buf = g_bytes_get_data(response_data, &response_len);
+			if (response_len > (gsize)FU_DEVLINK_NETLINK_BUF_SIZE) {
+				g_set_error(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INVALID_DATA,
+					    "recorded response too large: %zu > %ld",
+					    response_len,
+					    FU_DEVLINK_NETLINK_BUF_SIZE);
+				return FALSE;
+			}
+			memcpy(nlg->buf, response_buf, response_len); /* nocheck:blocked */
+			rc = response_len;
+		} else {
+			rc = mnl_socket_recvfrom(nlg->nl, nlg->buf, MNL_SOCKET_BUFFER_SIZE);
+			if (rc == 0) {
+				return TRUE;
+			}
+			if (rc < 0) {
+				g_set_error(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_NOT_SUPPORTED,
+					    "failed to receive netlink message: %s",
+					    fwupd_strerror(errno));
+				return FALSE;
+			}
+			if (nlg->save_events) {
+				g_autoptr(GBytes) response_data = g_bytes_new(nlg->buf, rc);
+				g_autofree gchar *response_key =
+				    g_strdup_printf("ResponseData%u", i++);
+
+				fu_device_event_set_bytes(event, response_key, response_data);
+			}
+		}
+
+		rc = fu_devlink_netlink_msg_cb_run(nlg, rc, seq, cb, data, error);
+	} while (rc > MNL_CB_STOP);
+
+	return rc == MNL_CB_STOP;
+}
+
+/* send message and receive response */
+gboolean
+fu_devlink_netlink_msg_send_recv(FuDevlinkGenSocket *nlg,
+				 struct nlmsghdr *nlh,
+				 mnl_cb_t cb,
+				 void *data,
+				 GError **error)
+{
+	gint rc;
+
+	if (!nlg->is_emulated) {
+		/* send netlink message */
+		rc = mnl_socket_sendto(nlg->nl, nlh, nlh->nlmsg_len);
+		if (rc < 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "failed to send netlink message: %s",
+				    fwupd_strerror(errno));
+			return FALSE;
+		}
+	}
+	return fu_devlink_netlink_msg_recv_run(nlg, nlh, cb, data, error);
+}
+
+gboolean
+fu_devlink_netlink_msg_send(FuDevlinkGenSocket *nlg, struct nlmsghdr *nlh, GError **error)
+{
+	return fu_devlink_netlink_msg_send_recv(nlg, nlh, NULL, NULL, error);
+}
+
+/* generic netlink control attribute callback */
+static int
+fu_devlink_netlink_genl_ctrl_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	if (mnl_attr_type_valid(attr, CTRL_ATTR_MAX) < 0)
+		return MNL_CB_OK;
+
+	tb[type] = attr;
```

Comment:
do we need to check type is < CTRL_ATTR_MAX? or is that what mnl_attr_type_valid does?

Reply:
EXPORT_SYMBOL int mnl_attr_type_valid(const struct nlattr *attr, uint16_t max)
{
        if (mnl_attr_get_type(attr) > max) {
                errno = EOPNOTSUPP;
                return -1;
        }
        return 1;
}

So we don't need to, mnl_attr_type_valid() does that.